x/bep3/keeper/swap.go

```diff
@@ -0,0 +1,340 @@
+	hashedSecret, err := types.CalculateSwapID(hashedSubmittedNumber, atomicSwap.From, atomicSwap.SenderOtherChain)
+	if err != nil {
+		return sdk.ErrInternal(err.Error())
+	}
+
+	if !bytes.Equal(hashedSecret, swapID) {
+		return types.ErrInvalidClaimSecret(k.codespace, hashedSecret, swapID)
+	}
+
+	// If AtomicSwap is not cross-chain, atomicSwap.ExpectedIncome equals atomicSwap.Amount
+	claimerCoins, err := sdk.ParseCoins(atomicSwap.ExpectedIncome)
+	if err != nil {
+		return sdk.ErrInternal(err.Error())
+	}
+
+	if atomicSwap.CrossChain {
+		err := k.ValidateCoinMint(ctx, atomicSwap.Amount)
+		if err != nil {
+			return err
+		}
+
+		// Mint full amount of this coin's associated debt coin to bep3 module for internal limit tracking
+		internalTrackingCoins, err2 := getEqualInternalTrackingCoins(atomicSwap.Amount)
+		if err2 != nil {
+			return sdk.ErrInternal(err2.Error())
+		}
+
+		err = k.supplyKeeper.MintCoins(ctx, types.ModuleName, internalTrackingCoins)
+		if err != nil {
+			return err
+		}
+
+		// Mint coins for distribution
+		err = k.supplyKeeper.MintCoins(ctx, types.ModuleName, atomicSwap.Amount)
+		if err != nil {
+			return err
+		}
+		// Send claimer their portion of coins
+		err = k.supplyKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, from, claimerCoins)
+		if err != nil {
+			return err
+		}
+		// Send deputy remaining coins
+		deputyCoins := atomicSwap.Amount.Sub(claimerCoins)
+		err = k.supplyKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, k.GetBnbDeputyAddress(ctx), deputyCoins)
+		if err != nil {
+			return err
+		}
+	} else {
+		// Send amount from bep3 module to adresss that successfully claimed AtomicSwap
+		err = k.supplyKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, from, claimerCoins)
+		if err != nil {
+			return sdk.ErrInternal(err.Error())
+		}
+	}
+
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			types.EventTypeClaimAtomicSwap,
+			sdk.NewAttribute(types.AttributeKeyAtomicSwapID, fmt.Sprintf("%s", atomicSwap.SwapID)),
+			sdk.NewAttribute(types.AttributeKeyClaimer, fmt.Sprintf("%s", from)),
+			sdk.NewAttribute(types.AttributeKeyCoinDenom, fmt.Sprintf("%s", claimerCoins[0].Denom)),
+			sdk.NewAttribute(types.AttributeKeyCoinAmount, fmt.Sprintf("%d", claimerCoins[0].Amount.Int64())),
+		),
+	)
+
+	// Update AtomicSwap state
+	k.DeleteAtomicSwap(ctx, swapID)
+
+	return nil
+}
+
+// RefundAtomicSwap refunds an AtomicSwap, sending assets to the original sender and closing the AtomicSwap
+func (k Keeper) RefundAtomicSwap(ctx sdk.Context, from sdk.AccAddress, swapID []byte) sdk.Error {
+
+	atomicSwap, found := k.GetAtomicSwap(ctx, swapID)
+	if !found {
+		return types.ErrAtomicSwapNotFound(k.codespace, swapID)
+	}
+
+	// Refund request must come from original creator or bep3 module
+	if !from.Equals(atomicSwap.From) && !k.GetBnbDeputyAddress(ctx).Equals(atomicSwap.From) {
+		return types.ErrOnlyOriginalCreator(k.codespace, from, atomicSwap.From)
+	}
+
+	if !atomicSwap.CrossChain {
```

Comment:
If a swap is cross chain and it's refunded, the only thing that happens is that the swap get's deleted. Is that correct?

Reply:
That was correct, but the new refactor now returns deposited coins to the original swap creator and updates the swap's status to 'Expired'.